typescript/agentkit/src/action-providers/compound/utils.ts

```diff
@@ -0,0 +1,515 @@
+  const borrowValue = borrowDetails.borrowAmount.mul(borrowDetails.price);
+  let totalAdjustedCollateral = new Decimal(0);
+
+  for (const supply of supplyDetails) {
+    const supplyTokenSymbol = supply.tokenSymbol;
+    const withdrawTokenSymbol = await getTokenSymbol(wallet, tokenAddress);
+
+    if (supplyTokenSymbol === withdrawTokenSymbol) {
+      const decimals = await getTokenDecimals(wallet, tokenAddress);
+      const withdrawAmountHuman = new Decimal(formatAmountFromDecimals(amount, decimals));
+      const newSupplyAmount = supply.supplyAmount.sub(withdrawAmountHuman);
+      const assetValue = newSupplyAmount.mul(supply.price);
+      totalAdjustedCollateral = totalAdjustedCollateral.add(
+        assetValue.mul(supply.collateralFactor),
+      );
+    } else {
+      totalAdjustedCollateral = totalAdjustedCollateral.add(
+        supply.supplyAmount.mul(supply.price).mul(supply.collateralFactor),
+      );
+    }
+  }
+
+  return borrowValue.eq(0) ? new Decimal(Infinity) : totalAdjustedCollateral.div(borrowValue);
+};
+
+/**
+ * Get health ratio after a hypothetical borrow
+ *
+ * @param wallet - The wallet provider instance
+ * @param cometAddress - The address of the Comet contract
+ * @param amount - The amount to borrow
+ * @returns The health ratio after borrow as a Decimal
+ */
+export const getHealthRatioAfterBorrow = async (
+  wallet: EvmWalletProvider,
+  cometAddress: `0x${string}`,
+  amount: bigint,
+): Promise<Decimal> => {
+  if (!cometAddress) {
+    throw new Error("Comet address is undefined");
+  }
+
+  const borrowDetails = await getBorrowDetails(wallet, cometAddress);
+  const supplyDetails = await getSupplyDetails(wallet, cometAddress);
+
+  const baseToken = (await wallet.readContract({
+    address: cometAddress,
+    abi: COMET_ABI,
+    functionName: "baseToken",
+    args: [],
+  })) as `0x${string}`;
+  const baseDecimals = await getTokenDecimals(wallet, baseToken);
+
+  const additionalBorrow = new Decimal(formatAmountFromDecimals(amount, baseDecimals));
+  const newBorrow = borrowDetails.borrowAmount.add(additionalBorrow);
+  const newBorrowValue = newBorrow.mul(borrowDetails.price);
+
+  let totalAdjustedCollateral = new Decimal(0);
+  for (const supply of supplyDetails) {
+    totalAdjustedCollateral = totalAdjustedCollateral.add(
+      supply.supplyAmount.mul(supply.price).mul(supply.collateralFactor),
+    );
+  }
+
+  return newBorrowValue.eq(0) ? new Decimal(Infinity) : totalAdjustedCollateral.div(newBorrowValue);
+};
+
+/**
+ * Get portfolio details in markdown format
+ *
+ * @param wallet - The wallet provider instance
+ * @param cometAddress - The address of the Comet contract
+ * @returns A markdown formatted string with portfolio details
+ */
+export const getPortfolioDetailsMarkdown = async (
+  wallet: EvmWalletProvider,
+  cometAddress: `0x${string}`,
+): Promise<string> => {
+  if (!cometAddress) {
+    throw new Error("Comet address is undefined");
+  }
```

Comment:
You can probably omit this check, since the type of `cometAddress` is not nullable or undefined. Up to you!